### PR TITLE
Adds election module example

### LINKED ISSUE #37: sov-modules: Example of how to implement a simple module
We need a simple reference implementation of a basic module.
It should show the following:

1. How to initialize the module.
2. Call a method and change the module's state.
3. Query the state of the module.

### PR DESCRIPTION
**Pull request by bkolad**
Example of a simple "election" module.

close #37

### CODE REVIEW

File: sov-modules/sov-modules-impl/src/example_election/call.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        match voter {
+            Voter::Voted => bail!("Voter tried voting a second time!"),
+            Voter::Fresh => {
+                self.allowed_voters.set(context.sender(), Voter::voted());
+
+                let mut candidates = self.get_candidates()?;
+
+                // Check if a candidate exist.
+                let mut candidate = candidates
+                    .get(candidate_index)
```
[the remark this answers is not in the document]
Comment (bkolad) [reply sov-modules/sov-modules-impl/src/example_election/call.rs]: Yes! Done.

File: sov-modules/sov-modules-impl/src/example_election/call.rs [hunk 2]
```diff
@@ -74,17 +74,16 @@ impl<C: sov_modules_api::Context> Election<C> {
...
+                    .ok_or(anyhow!("Candidate doesn't exist"))?;
 
                 candidate.count = candidate
                     .count
                     .checked_add(1)
                     .ok_or(anyhow!("Vote count overflow"))?;
 
                 // We already checked that the candidate_index exists.
-                candidates[candidate_index] = candidate;
+                candidates[candidate_index] = candidate.clone();
```
Comment (preston-evans98) [inline sov-modules/sov-modules-impl/src/example_election/call.rs:86]: Is this needed now? Shouldnt we be able to just remove lines 85-86?

File: sov-modules/sov-state/src/internal_cache.rs [hunk 3]
```diff
@@ -39,7 +39,19 @@ impl StorageInternalCache {
                 cache_value_exists.value.map(|value| StorageValue { value })
             }
             // If the value does not exist in the cache, then fetch it from an external source.
-            cache::ValueExists::No => value_reader.read_value(key),
+            cache::ValueExists::No => {
+                let value = value_reader.read_value(key);
+
+                let cache_value = CacheValue {
+                    value: value.clone().map(|storage_value| storage_value.value),
```
[the remark this answers is not in the document]
Comment (bkolad) [reply sov-modules/sov-state/src/internal_cache.rs]: Fixed

File: sov-modules/sov-modules-impl/src/example_election/query.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+impl<C: sov_modules_api::Context> Election<C> {
+    pub fn results(&self) -> Response {
+        let is_frozen = self.is_frozen.get().unwrap_or_default();
+
+        if is_frozen {
+            let candidates = self.candidates.get().unwrap_or(Vec::default());
+
+            // In case of tie, returns the candidate with the higher index in the vec.
```
[the remark this answers is not in the document]
Comment (bkolad) [reply sov-modules/sov-modules-impl/src/example_election/query.rs]: Yeah I have added a short sentence at the end.

File: sov-modules/sov-modules-impl/src/example_election/call.rs (hunk 2 above)
Comment (bkolad) [reply sov-modules/sov-modules-impl/src/example_election/call.rs:86]: Yes so `candidate` is a reference but the `candidates` vec keeps owned values.

File: sov-modules/sov-modules-impl/src/example_election/call.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            .allowed_voters
+            .get(context.sender())
+            .ok_or(anyhow!("Voter missing from the allowed list."))?;
+
+        match voter {
+            Voter::Voted => bail!("Voter tried voting a second time!"),
+            Voter::Fresh => {
+                self.allowed_voters.set(context.sender(), Voter::voted());
+
+                let mut candidates = self.get_candidates()?;
```
[the remark this answers is not in the document]
Comment (bkolad) [reply sov-modules/sov-modules-impl/src/example_election/call.rs:73]: I will think about it and create an issue.

File: sov-modules/sov-modules-impl/src/example_election/call.rs (hunk 2 above)
Comment (preston-evans98) [reply sov-modules/sov-modules-impl/src/example_election/call.rs:86]: I must be missing something - but can't we mutate the value in place through the mutable reference? If `candidate` is a `&mut Candidates[idx]`, doesn't `candidate.count = ...` already change the value in `candidates`?

Comment (bkolad) [reply sov-modules/sov-modules-impl/src/example_election/call.rs]: Of course we can I don't know why I didn't see it! Fixed.

File: sov-modules/sov-modules-impl/src/example_election/call.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (preston-evans98) [reply sov-modules/sov-modules-impl/src/example_election/call.rs:73]: Sounds great, thanks!

**Review by preston-evans98 — APPROVED** [review]
Looks good!

**Pull request merged**